incydr/cli/cmds/users.py

```diff
@@ -239,33 +270,135 @@ def move(ctx: Context, user, org_guid):
+@single_format_option
+@click.pass_context
+def list_roles(ctx: Context, format_: SingleFormat = None):
+    """
+    List all available roles that can be assigned by the current user.
+    """
+    client = ctx.obj()
+    roles = client.users.v1.list_roles()
+    if format_ == TableFormat.csv:
+        render.csv(Role, roles, flat=True)
+    elif format_ == TableFormat.table:
+        render.table(Role, roles, flat=False)
+    elif format_ == TableFormat.json:
+        for item in roles:
+            console.print_json(item.json())
+    else:  # raw-json
+        for item in roles:
+            click.echo(item.json())
+
+
 @users.command(cls=IncydrCommand)
-@click.argument("csv")
+@click.argument("file", type=AutoDecodedFile())
+@input_format_option
+@click.option("--add", "update_method", flag_value="add", default=None)
+@click.option("--remove", "update_method", flag_value="remove", default=None)
 @click.pass_context
-# TODO: role IDs
-def bulk_update_roles(ctx: Context, csv):
+def bulk_update_roles(ctx: Context, file, update_method=None, format_=None):
     """
     Bulk update roles associated with multiple users with a CSV file.
 
+     By default, the provided roles will replace the specified user's existing roles.
+     Use the --add flag or the --remove flag to add or remove roles, respectively, from a user's existing roles.
+
     Takes a single arg `CSV` which specifies the path to the file.
     Requires the following columns:
 
     * `user` - User ID or username of the user whose roles will be updated. Performs an additional lookup if username is passed.
-    * `roles` - Space-delimited list of role IDs to assign to the new user.  These will replace the specified user's existing roles.
+    * `role` - Role ID and/or role name (case-sensitive) to assign to the new user.
     """
+
+    @lru_cache()
+    def resolve_username(user):
+        if user is None:
+            return
+        elif "@" in user:
+            return user_lookup(client, user)
+        else:  # assume user_id
+            return user
+
+    class RoleUpdateCSV(CSVModel):
+        user: str = Field(csv_aliases=["user"])
+        role: str = Field(csv_aliases=["role"])
+
+    class RoleUpdateJSON(Model):
+        user: str
+        role: str
+
+    if format_ == "csv":
+        roles_file = RoleUpdateCSV.parse_csv(file)
+    else:
+        roles_file = RoleUpdateJSON.parse_json_lines(file)
+
     client = ctx.obj()
-    try:
-        for row in track(
-            read_dict_from_csv(csv),
-            description="Updated user roles...",
-            transient=True,
-        ):
-            user = row.get("user")
-            if user and "@" in user:
-                user = user_lookup(client, row["user"])
-            client.users.v1.update_roles(user, row["roles"].split(" "))
-    except HTTPError as err:
-        console.print(f"[red]Error:[/red] {err.response.text}")
+    if update_method == "add":
+
+        def update_user_roles(user_, roles):
```

Comment:
A new function doesn't need to be defined for each of these. Since the client methods all take the same args, we can just assign the existing method we want to a variable:

```
if update_method == "add":
    update_func = client.users.v1.add_roles
elif update_method == "remove":
    update_func = client.users.v1.remove_roles
...
```

Reply:
oohh thats right, python function-as-object magic 🪄